src/libpiano/response.c

```diff
@@ -241,6 +316,173 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+							break;
+						}
+						assert (trackInfo!= NULL);
+						song->artist = PianoJsonStrdup(trackInfo, "artistName");
+						song->album = PianoJsonStrdup(trackInfo, "albumName");
+						song->title = PianoJsonStrdup(trackInfo, "name");
+						song->fileGain = 0.0;
+						song->length = getInt(trackInfo, "duration");
+						song->coverArt = getCoverArt(trackInfo);
+						playlist = PianoListAppendP (playlist, song);
+					}
+					break;
+				}
+
+				case PIANO_TYPE_ALBUM: {
+					char trackTitle[120];
+					int totalTracks = 0;
+					int trackNumber;
+
+				// Count tracks
+					json_object_object_foreach(result,Key1,Val1) {
+						if(Key1[0] != 'T' || Key1[1] != 'R') {
+							continue;
+						}
+						totalTracks++;
+					}
+
+					json_object_object_foreach(result,Key,Val) {
+						if(Key[0] != 'T' || Key[1] != 'R') {
+							continue;
+						}
+						LOG("got track %s: ",Key);
+						trackNumber = getInt(Val,"trackNumber");
+						snprintf(trackTitle,sizeof(trackTitle),
+									totalTracks > 9 ? "%02d %s" : "%d %s",
+									trackNumber,PianoJsonGetStr(Val,"name"));
+						LOG_RAW("%s\n",trackTitle);
+						if(getBool(Val,"/rightsInfo/hasInteractive") != 1) {
+							LOG(" ignored\n");
+							LOG("  hasInteractive: %d\n",getBool(Val,"/rightsInfo/hasInteractive"));
+							LOG("  hasOffline: %d\n",getBool(Val,"/rightsInfo/hasOffline"));
+							LOG("  hasNonInteractive: %d\n",getBool(Val,"/rightsInfo/hasNonInteractive"));
+							LOG("  hasStatutory: %d\n",getBool(Val,"/rightsInfo/hasStatutory"));
+							LOG("  hasRadioRights: %d\n",getBool(Val,"/rightsInfo/hasRadioRights"));
+							continue;
+						}
+						PianoSong_t *song;
+
+						if ((song = calloc (1, sizeof (*song))) == NULL) {
+							return PIANO_RET_OUT_OF_MEMORY;
+						}
+
+						song->stationId = strdup(reqData->station->id);
+						song->title = strdup(trackTitle);
+						song->trackToken = strdup(Key);
+						song->seedId = PianoJsonStrdup(Val,"albumId");
+						song->artist = PianoJsonStrdup(Val,"artistName");
+						song->album = PianoJsonStrdup(Val,"albumName");
+						song->fileGain = 0.0;
+						song->length = getInt(Val, "duration");
+						song->coverArt = getCoverArt(Val);
+					// Add to playlist in track order
+
+						if(playlist == NULL) {
+							playlist = song;
+						}
+						else {
+							PianoSong_t *lastSong = (PianoSong_t *) &playlist;
+							PianoSong_t *nextSong = playlist;
+							do {
+								int nextSongTrackNum;
+								if(nextSong == NULL) {
+									lastSong->head.next = (struct PianoListHead *) song;
+									break;
+								}
+								sscanf(nextSong->title,"%d",&nextSongTrackNum);
+								if(nextSongTrackNum > trackNumber) {
+									lastSong->head.next = (struct PianoListHead *) song;
+									song->head.next = (struct PianoListHead *) nextSong;
+									break;
+								}
+								lastSong = nextSong;
+								nextSong = (PianoSong_t *) nextSong->head.next;
+							} while(true);
```

Comment:
So, basically, the result returned by Pandora is not sorted and we need to do that here?